docs/reddit-migration/strategy/schema.md

```diff
@@ -0,0 +1,99 @@
+ - `text` / `preview_text` / `plain_text`
+   - We currently have several representations of textual post content, the 3 fields/properties listed above:
+     - `text` is the markdown text for text posts
+     - `preview_text` is precomputed plain text representation of link posts, sourced from and Embedly API request for that link
+     `plain_text` is a `@property`, not a db column, but its logic is as follows:
+       - If the post is an article, render the article to plain text
+       - If the post is a text post, render the markdown to plain text
+       - Otherwise use `preview_text`
+    - See also: https://github.com/mitodl/open-discussions/blob/9a96f09ab4c8c617da3ba8842adb9c419c5d321d/docs/rfcs/0002-indexing-embedly-link-preview.md#4-related-but-separate-issue-getting-rid-of-either-the-plain_text-or-text-field-in-es-documents
+    - We should to normalize `preview_text` / `plain_text` down to a single stored field so we're not recomputing it every time a read occurs.
+    - We still need to keep the `text` field and the `article` relation
+
+
+### Comments
+
+Comments should have the fields carried over with the following adjustments:
+
+ - `channel_id` -> `id`
+ - `reports_ignored = BooleanField(default=False)`
+   - Flag to indicate whether we're ignoring subsequent reports of the comment
+   - See "Moderation" for more details
+
+#### Comment Tree
+
+Prior art:
+
+ - SQL Antipatterns ("Naive Trees"):http://www.r-5.org/files/books/computers/languages/sql/style/Bill_Karwin-SQL_Antipatterns-EN.pdf
+ - Models for Hierarchical Data: https://www.slideshare.net/billkarwin/models-for-hierarchical-data
+ - Django treebeard: https://django-treebeard.readthedocs.io/en/latest/
+
+Reddit stores comment trees in Cassandra, which is overkill for our current use cases. That said, we need to be able to migrate the comments in such a manner that we achieve the following:
+
+- Maintain current API result structures
+  - Specifically, we need to support some form of reddit's "more comments" pagination. This **does not** need to partition the comments exactly as reddit would, the interface just needs to be compatible with how the frontend traverses it.
+- Support currently used sorting methods (best, old, new)
+- Support rendering an arbitrary subtree of the comments for the comment detail view
+
+From the Prior Art listed above, Materialized Paths are probably our best bet and we're already familiar to some degree with them because they're what `wagtail` uses under the hood for page hierarchies.
+
+The big thing here is that in order to support the sorts we need one for *each type of sort*. That precludes using a single table for everything so we'll want to have a hierarchical table for each sort and then `prefetch_related` on the `Comment` table so that code can work generically over similarly structured tables. This approach means these sorting hierarchy tables end up being a sorted lookup table for extracting a portion of the comment tree.
```

Comment:
does "one" in the first sentence in this paragraph mean that we need a separate materialized path for each sort? that's what I take this to mean but just wanted to confirm.

that seems like a decent way to do it to me. Django-treebeard supports it as well. 

By having a hierarchical table for each sort, do you mean (to use some terminology from the treebeard project) we would have something like

```py
class BestComment(MP_Node):
    comment = models.ForeignKey(Comment)
    post = models.ForeignKey(post)

class OldComment(MP_Node):
    comment = models.ForeignKey(Comment)
    post = models.ForeignKey(post)

class NewComment(MP_Node):
    comment = models.ForeignKey(Comment)
    post = models.ForeignKey(post)
```

or something along those lines? it seems like it could get tricky to update all of those and keep them in sync. I guess in my pseudocode I assumed that the comment text, votes, etc would be stored in a separate table so it's not duplicated. I guess this reason is why reddit comments still show up in the tree after they have been deleted?

My question here would be to ask whether we're over complicating by assuming we want to support all of these sorts. Are we sure this is something we want to emulate, or would we be satisfied with only one of the three?

It seems to me that the 'old' and 'new' ones are the simplest, since they are simply adding new entries to the tree. but the 'best' one would have to be recalculated on every vote, and potentially a whole series of adjustments made. I guess we could do this intelligently - like if there's a vote on a top-level comment we're only going to be making adjustments to the top-level, not to the whole tree, and the only possible adjustment would be if the score of one comment at a given level in the hierarchy is now higher or lower than another they have to be reordered. But idk, I'm wondering about the performance / reliability aspects of trying to do something like that efficiently and reliably when a lot of upvotes are coming in (not that we see much traffic on production now but still).

Reply:
Yeah, that's generally the data model we'd end up using, I have a functional branch here: https://github.com/mitodl/open-discussions/tree/nl/reddit_schema_migration

Materialized paths are pretty efficient with moves/reordering. Treebeard abstracts that all away from us so it ends up only being a few lines of code. Here's how you'd do that with my branch:

```python
from django.db import transaction
from discussions.models import *
channel = Channel.objects.create()
post = Post.objects.create(channel=channel)
comment1 = Comment.objects.create(post=post)
comment2 = Comment.objects.create(post=post)
comment3 = Comment.objects.create(post=post)

@transaction.atomic
def rescore(node, new_score):
    node.score = score
    node.save()
    node.move(node)

# comment trees start with a root node w/ `comment=None`
# this would get created when the post is created
root = BestCommentTreeNode.add_root(post=post, score=0) 
# top level comment
c1n = root.add_child(comment=comment1, score=1)
# replies to comment1
c2n = c1n.add_child(comment=comment1, score=4)
c3n = c1n.add_child(comment=comment1, score=5)

# c3n should list above c2n
print(root.dump_bulk())

rescore(c2n, 8)
# c2n is now sorted above c3n and has the updated score
print(root.dump_bulk())
```

The 4 queries the move generates are (fair warning it's a bit noisy, but it's 2 `UPDATE`s and 2 `SELECT LIMIT 1`s):

```
UPDATE "discussions_bestcommenttreenode" SET "path" = '000100010002', "depth" = 3, "numchild" = 0, "post_id" = 2, "comment_id" = 3, "score" = 8 WHERE "discussions_bestcommenttreenode"."id" = 5;
SELECT "discussions_bestcommenttreenode"."id", "discussions_bestcommenttreenode"."path", "discussions_bestcommenttreenode"."depth", "discussions_bestcommenttreenode"."numchild", "discussions_bestcommenttreenode"."post_id", "discussions_bestcommenttreenode"."comment_id", "discussions_bestcommenttreenode"."score" FROM "discussions_bestcommenttreenode" WHERE ("discussions_bestcommenttreenode"."depth" = 3 AND "discussions_bestcommenttreenode"."path" BETWEEN '000100010000' AND '00010001zzzz' AND "discussions_bestcommenttreenode"."score" > 8) ORDER BY "discussions_bestcommenttreenode"."path" ASC  LIMIT 1; 
SELECT "discussions_bestcommenttreenode"."id", "discussions_bestcommenttreenode"."path", "discussions_bestcommenttreenode"."depth", "discussions_bestcommenttreenode"."numchild", "discussions_bestcommenttreenode"."post_id", "discussions_bestcommenttreenode"."comment_id", "discussions_bestcommenttreenode"."score" FROM "discussions_bestcommenttreenode" WHERE ("discussions_bestcommenttreenode"."depth" = 3 AND "discussions_bestcommenttreenode"."path" BETWEEN '000100010000' AND '00010001zzzz') ORDER BY "discussions_bestcommenttreenode"."path" DESC  LIMIT 1;
UPDATE "discussions_bestcommenttreenode" SET path='000100010005'||SUBSTR(path, 13) WHERE path LIKE '000100010002%';
```

Reply:
that doesn't look too bad at all. has your experience been that the library is pretty nice to work with? I am no data structures expert but based on a little reading it does seem like a materialized path is the most straightforward option, and this package seems to have a good implementation of it.

does it handle removal / deletion of nodes that have children? it seems that the leaf node case is easy to handle.

Reply:
Yeah, it handles all of that pretty easily, I'm not sure in practice if we'll use the entirety of the possible operations, but it's very easy to work with for the use cases we have.